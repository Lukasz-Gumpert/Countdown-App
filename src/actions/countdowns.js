import uuid from 'uuid';
import database from '../firebase/firebase';

export const addCountdown = (countdown) => ({
  type: 'ADD_COUNTDOWN',
  countdown
});

export const startAddCountdown = (countdown = {}) => {
  return (dispatch) => {
    const {
      title = '',
      category = '',
      releaseDate = 0,
      description = '',
    } = countdown;
    const countdownData = { title, category, releaseDate, description };
    return database.ref('countdowns').push(countdownData).then((ref) => {
      dispatch(addCountdown({
        id: ref.key,
        ...countdownData
      }));
    });
  };
};

export const removeCountdown = ({ id }) => ({
  type: 'REMOVE_COUNTDOWN',
  id
});

export const editCountdown = (id, updates) => ({
  type: 'EDIT_COUNTDOWN',
  id,
  updates
});

export const setCountdowns = (countdowns) => ({
  type: 'SET_COUNTDOWNS',
  countdown
})