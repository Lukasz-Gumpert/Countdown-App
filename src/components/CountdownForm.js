import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

export default class CountdownForm extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            title: '',
            category: 'movies',
            releaseDate: moment(),
            description: '',
            calendarFocused: false,
            error: '',
        };
    };

    onTitleChange = (e) => {
        const title = e.target.value;
        this.setState(() => ({ title }));
        console.log(this.state.title)
    };

    onCategoryChange = (sel) => {
        const category = sel.target.value;
        this.setState(() => ({ category }));
    };

    onDateChange = (releaseDate) => {
        if (releaseDate) {
            this.setState(() => ({ releaseDate }));
        };
    };

    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }));
    };

    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({ description }))
    };

    onSubmit = (e) => {
        e.preventDefault();
        this.props.onSubmit({
            title: this.state.title,
            category: this.state.category,
            releaseDate: this.state.releaseDate.valueOf(),
            description: this.state.description
        });
    };

    render() {
        return (
            <form  onSubmit={this.onSubmit}>
                <input
                    className="text-input"
                    type="text"
                    placeholder="title"
                    autoFocus
                    value={this.state.title}
                    onChange={this.onTitleChange}
                />
                <select
                    value={this.state.category}
                    onChange={this.onCategoryChange}
                    name="category"
                    className="select-category"
                >
                    <option value="movies">Movies</option>
                    <option value="music">Music</option>
                    <option value="tech">Tech</option>
                    <option value="games">Games</option>
                    <option value="events">Events</option>
                </select>
                <SingleDatePicker
                    date={this.state.releaseDate}
                    onDateChange={this.onDateChange}
                    focused={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                />
                <input
                    className="text-input"
                    type="text"
                    placeholder="description"
                    value={this.state.description}
                    onChange={this.onDescriptionChange}
                />
                <button>Save</button>
            </form>

        )
    }
}