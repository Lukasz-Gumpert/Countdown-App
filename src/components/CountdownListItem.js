import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment'
import CountdownClock from './CountdownClock';

const CountdownListItem = ({ id, title, category, releaseDate, description }) => (
    <Link to={`edit/${id}`} className="list-item">
        <p> {category} </p>
        <h1>{title}</h1> : {description} <p>Time left:</p>
        <CountdownClock timeTillDate={moment(releaseDate)} />
    </Link>
);

export default CountdownListItem;