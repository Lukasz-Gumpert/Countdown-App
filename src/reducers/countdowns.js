const countdownsReducerDefaultState = [];

export default (state = countdownsReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_COUNTDOWN':
            return [
                ...state,
                action.countdown
            ];
        case 'EDIT_COUNTDOWN':
            return state.map((countdown) => {
                if (countdown.id === action.id) {
                    return {
                        ...countdown,
                        ...action.updates
                    }
                } else {
                    return countdown;
                };
            });
        case 'REMOVE_COUNTDOWN':
            return state.filter(({ id }) => {
                return id !== action.id
            });
        case 'SET_COUNTDOWNS':
            return action.countdowns;
        default:
            return state;
    };
};