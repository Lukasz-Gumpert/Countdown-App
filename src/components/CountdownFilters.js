import React from 'react';
import { connect } from 'react-redux';
import { setTextFilter, sortByName, sortByType, endingSoonest, furthestAway } from '../actions/filters';

export class CountdownFilters extends React.Component {
    constructor(props) {
        super(props);

        this.onTextChange = (e) => {
            return this.props.setTextFilter(e.target.value);
        }

        this.onSortChange = (sel) => {
            const sortBy = sel.target.value;
            if (sortBy === "sortByName") {
                return props.sortByName();
            } else if (sortBy === "sortByType") {
                return props.sortByType();
            } else if (sortBy === "endingSoonest") {
                return props.endingSoonest();
            } else if (sortBy === "furthestAway") {
                return props.furthestAway();
            };
        };
    };

    render() {
        return (
            <div>
                <input
                    type="text"
                    placeholder="search"
                    value={this.props.filters.text}
                    onChange={this.onTextChange}
                />
                <select
                    value={this.props.filters.sortBy}
                    onChange={this.onSortChange}
                    name="sortBy"
                >
                    <option value="endingSoonest">Ending soonest</option>
                    <option value="furthestAway" >Furthest away</option>
                    <option value="sortByName" >By name</option>
                    <option value="sortByType" >By type</option>
                </select>
            </div>
        )
    }
}

const mapStateToProps = (state) => {
    return {
        filters: state.filters
    };
};

const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (data) => dispatch(setTextFilter(data)),
    sortByName: () => dispatch(sortByName()),
    sortByType: () => dispatch(sortByType()),
    endingSoonest: () => dispatch(endingSoonest()),
    furthestAway: () => dispatch(furthestAway()),
});


export default connect(mapStateToProps, mapDispatchToProps)(CountdownFilters);