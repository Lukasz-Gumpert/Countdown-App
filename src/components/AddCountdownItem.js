import React from 'react';
import { connect } from 'react-redux';
import CountdownForm from './CountdownForm';
import { startAddCountdown } from '../actions/countdowns';
import moment from 'moment';

export class AddCountdownItem extends React.Component {
    onSubmit = (countdown) => {
        this.props.startAddCountdown(countdown);
        this.props.history.push('/');

    }
 
    render() {
        return (
            <div>
                <h1>Add countdown</h1>
                <CountdownForm onSubmit={this.onSubmit} />
            </div>
        )
    }
}

const mapDispatchToProps = (dispatch) => ({
    startAddCountdown: (countdown) => dispatch(startAddCountdown(countdown))
});

export default connect(undefined, mapDispatchToProps)(AddCountdownItem);
