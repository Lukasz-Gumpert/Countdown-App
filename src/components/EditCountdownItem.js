import React from 'react';
import { connect } from 'react-redux';
import CountdownForm from './CountdownForm';
import { editCountdown, removeCountdown } from '../actions/countdowns';

export class EditCountdownItem extends React.Component {

    onSubmit = (countdown) => {
        this.props.editCountdown(this.props.countdown.id, countdown);
        this.props.history.push('/');
    }
    onRemove = () => {
        this.props.removeCountdown(this.props.countdown);
        this.props.history.push('/');
    }
    render() {
        return (
            <div>
                <p>Edit</p>
                <CountdownForm
                    countdown={this.props.countdown}
                    onSubmit={this.onSubmit}
                />
                <div>
                    <button onClick={this.onRemove}>Remove</button>
                </div>
            </div>
        )
    }
};

const mapStateToProps = (state, props) => {
    return {
        countdown: state.countdowns.find((countdown) => countdown.id === props.match.params.id)
    }
};

const mapDispatchToProps = (dispatch, props) => {
    return {
        editCountdown: (id, countdown) => dispatch(editCountdown(id, countdown)),
        removeCountdown: (countdown) => dispatch(removeCountdown(countdown))
    }
};

export default connect(mapStateToProps, mapDispatchToProps)(EditCountdownItem);