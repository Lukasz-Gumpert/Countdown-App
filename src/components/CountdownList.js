import React from 'react';
import { connect } from "react-redux";
import CountdownListItem from '../components/CountdownListItem';
import selectFilters from '../selectors/filters';


export const CountdownList = (props) => (
    <div>
        {
            props.countdowns.length === 0 ? (
                <p>No countdown data</p>
            ) : (
                    props.countdowns.map((countdown) => {
                        return <CountdownListItem key={countdown.id} {...countdown} />
                    })
                )
        }
    </div>
);

const mapStateToProps = (state) => {
    return {
        countdowns: selectFilters(state.countdowns, state.filters)
    }
};

export default connect(mapStateToProps)(CountdownList);