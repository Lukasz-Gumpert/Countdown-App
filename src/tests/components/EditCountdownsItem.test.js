import React from 'react';
import { shallow } from 'enzyme';
import { EditCountdownItem } from '../../components/EditCountdownItem';
import countdowns from '../fixtures/countdowns';

let wrapper, editCountdown, removeCountdown, history

beforeEach(() => {
    editCountdown = jest.fn();
    removeCountdown = jest.fn();
    history = { push: jest.fn() };
    wrapper = shallow(
        <EditCountdownItem
            editCountdown={editCountdown}
            removeCountdown={removeCountdown}
            history={history}
            countdown={countdowns[3]}
        />);
});

test('Should render EditCountdownItem correctly', () => {
expect(wrapper).toMatchSnapshot();
});

test('Should handle editCountdown', () => {
wrapper.find('CountdownForm').prop('onSubmit')(countdowns[3]);
expect(history.push).toHaveBeenLastCalledWith('/');
expect(editCountdown).toHaveBeenLastCalledWith(countdowns[3].id, countdowns[3]);
});

test('Should handle removeCountdown', () => {
wrapper.find('button').simulate('click');
expect(history.push).toHaveBeenLastCalledWith('/');
expect(removeCountdown).toHaveBeenLastCalledWith(countdowns[3]);
});