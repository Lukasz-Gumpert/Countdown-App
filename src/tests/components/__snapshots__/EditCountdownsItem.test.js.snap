// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render EditCountdownItem correctly 1`] = `
<div>
  <p>
    Edit
  </p>
  <CountdownForm
    countdown={
      Object {
        "category": "music",
        "description": "cd",
        "id": 4,
        "releaseDate": 1296000000,
        "title": "Chill mix vol 2",
      }
    }
    onSubmit={[Function]}
  />
  <div>
    <button
      onClick={[Function]}
    >
      Remove
    </button>
  </div>
</div>
`;
